### PR TITLE
compile check to ensure cloud providers implement cloud interfaces

### PR DESCRIPTION
**Pull request by andrewsykim**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Ensures all providers are implementing cloud providers interfaces at compile time. As discussed in today's SIG meeting. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

```release-note
NONE
```

/sig cloud-provider
/assign @cheftako @justinsb

### CODE REVIEW

File: pkg/cloudprovider/providers/gce/gce.go [hunk 1]
```diff
@@ -98,6 +98,13 @@ type gceObject interface {
 	MarshalJSON() ([]byte, error)
 }
 
+var _ cloudprovider.Interface = (*Cloud)(nil)
+var _ cloudprovider.Instances = (*Cloud)(nil)
+var _ cloudprovider.LoadBalancer = (*Cloud)(nil)
+var _ cloudprovider.Routes = (*Cloud)(nil)
+var _ cloudprovider.Zones = (*Cloud)(nil)
+var _ cloudprovider.PVLabeler = (*Cloud)(nil)
```
Comment (cheftako) [inline pkg/cloudprovider/providers/gce/gce.go]: Missing "var _ cloudprovider.Clusters = (*Cloud)(nil)

**Review by cheftako — COMMENTED** [review]
We also have several package private objects which implement some of these interfaces. (Eg. azure_vmss scaleSet implements LoadBalancer) Wondering if it would be would be worth also adding checks for the package private objects.

**Comment by andrewsykim** [comment]
Good point @cheftako. I'll dig into that, thanks!

**Comment by andrewsykim** [comment]
@cheftako only internal type I could find was CloudStack's metadata (now updated). Couldn't find the azure_vmss scale set type you were referring to (though it does implement similar sounding methods like `EnsureHostsInPool`).

**Comment by cheftako** [comment]
> Walter Fender only internal type I could find was CloudStack's metadata (now updated). Couldn't find the azure_vmss scale set type you were referring to (though it does implement similar sounding methods like `EnsureHostsInPool`).

I believe azure_vmss scaleSet ends up "implmenting" LoadBalancer and other methods via composition with the Azure Cloud object. (https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/azure/azure_vmss.go#L51) 
This gets a bit nasty because scaleSet could (it doesn't) override one of the LoadBalancer methods. 
Warning by both scaleSet and Cloud failing to compile might be a hint.
Unfortunately when we fix Cloud it would also appear to fix scaleSet (however the old over-ride will now no longer over-ride or be called)

**Pull request merged**